### DISCUSSION TITLE
Schedule manager

### DISCUSSION THREAD

**Discussion by kotope** [jagheterfredrik/wallbox-mqtt-bridge#47] [Ideas]
Hi!

I've started looking into managing schedules locally. I require this functionality to actually automatically schedule my car charging for the cheapest electricity spot prices.

When updating database values of the scheudules table, also updating ulid and created_at fields, the data does not seem to sync to the Wallbox cloud.
Do you happen to have any idea which service is responsible of syncing the cloud and local DB? Or have you reverse-engineered it to see how is the sync being done?

**Comment by jagheterfredrik** [comment]
It’s totally doable but I would recommend using HA scheduler instead: https://github.com/nielsfaber/scheduler-card

**Reply by kotope** [reply]
Yes, I know I could already implement it with HA scheduler. :-)
..but I want to make it less dependant of the HA functionality: set the schedules straight to the Wallbox and let it handle the rest independently of Home Assistant.

Maybe it's a bit over-engineering, but I've been using same approach with the Shellys when scheduling heating and I've liked it so far. If the HA fails, I still have the old schedules active to start heating at proper times (in this case it would just be to charge the car).

Btw, thanks for publishing the root support & mqtt gateway!

**Reply by jagheterfredrik** [reply]
I see. I’m pretty sure the internal scheduler wants posix queue messages. The easiest way to understand their format is to strace the blewallbox process while modifying schedules, looking specifically for mq_timedsend syscall.

**Reply by kotope** [reply]
Ok. thanks for the tip! I'll check the posix queue messages next!

**Reply by jagheterfredrik** [reply]
I reversed the schedule format used over Bluetooth, chances are its propagated as is over the queue. I’ll dig up my notes

**Reply by jagheterfredrik** [reply]
Found it, first line is what is sent over BLE. Keep `type` at zero, I think.
```
# EaE.{"met":"w_sch","par":"1210022000650","id":571}
# 1 : schedule id
# 2100 : utc start
# 2200 : utc end
# 065 : days bitmask (sunday first bit, sat last)
# 0 : type (tried 2, got stuck, had to restore)
```

**Reply by jagheterfredrik** [reply]
It's apparently not using posix queues, it's using Redis PubSub, because why not. Here I'm setting 21-22 GMT+1, first all days, then then weekdays, then monday only, then removing the schedule again. Notice the CMD_READ_ALL_SCHEDULES in between, response is received on the `/wbx/schedule_manager/responses` channel.

```
# redis-cli
127.0.0.1:6379> subscribe /wbx/schedule_manager/cmd
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"body\":{\"schedule\":{\"created_at\":1707250056,\"enable\":true,\"enabled_days\":[],\"friday\":true,\"max_current\":1,\"monday\":true,\"name\":\"\",\"saturday\":true,\"schedule_id\":1,\"start_time\":2100,\"stop_time\":2200,\"sunday\":true,\"thursday\":true,\"tuesday\":true,\"type\":\"Regular\",\"ulid\":\"01HP00GZG3B6N5KPETZ5RCQTJ8\",\"wednesday\":true}},\"header\":{\"message_id\":\"CMD_UPDATE_SCHEDULE\",\"request_id\":\"MYWALLBOX#27\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:07:38.194Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"header\":{\"message_id\":\"CMD_READ_ALL_SCHEDULES\",\"request_id\":\"MYWALLBOX#28\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:07:51.242Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"body\":{\"schedule\":{\"created_at\":1707250069,\"enable\":true,\"enabled_days\":[],\"friday\":true,\"max_current\":1,\"monday\":true,\"name\":\"\",\"saturday\":false,\"schedule_id\":1,\"start_time\":2100,\"stop_time\":2200,\"sunday\":false,\"thursday\":true,\"tuesday\":true,\"type\":\"Regular\",\"ulid\":\"01HP00HC7PAJA4F7QNNRBA80KB\",\"wednesday\":true}},\"header\":{\"message_id\":\"CMD_UPDATE_SCHEDULE\",\"request_id\":\"MYWALLBOX#29\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:07:51.254Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"header\":{\"message_id\":\"CMD_READ_ALL_SCHEDULES\",\"request_id\":\"MYWALLBOX#30\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:08:22.628Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"body\":{\"schedule\":{\"created_at\":1707250101,\"enable\":true,\"enabled_days\":[],\"friday\":false,\"max_current\":1,\"monday\":true,\"name\":\"\",\"saturday\":false,\"schedule_id\":1,\"start_time\":2100,\"stop_time\":2200,\"sunday\":false,\"thursday\":false,\"tuesday\":false,\"type\":\"Regular\",\"ulid\":\"01HP00JB2QJ36C2PNKC30R3QXN\",\"wednesday\":false}},\"header\":{\"message_id\":\"CMD_UPDATE_SCHEDULE\",\"request_id\":\"MYWALLBOX#31\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:08:22.638Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"header\":{\"message_id\":\"CMD_READ_ALL_SCHEDULES\",\"request_id\":\"MYWALLBOX#32\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:08:28.749Z\"}}"
1) "message"
2) "/wbx/schedule_manager/cmd"
3) "{\"body\":{\"schedule\":{\"created_at\":1707250101,\"enable\":false,\"enabled_days\":[false,true,false,false,false,false,false],\"friday\":false,\"max_current\":1,\"monday\":true,\"name\":\"\",\"saturday\":false,\"schedule_id\":1,\"start_time\":2100,\"stop_time\":2200,\"sunday\":false,\"thursday\":false,\"tuesday\":false,\"type\":\"Regular\",\"ulid\":\"01HP00JJANKC9BF6A5DRNNHF2K\",\"wednesday\":false}},\"header\":{\"message_id\":\"CMD_UPDATE_SCHEDULE\",\"request_id\":\"MYWALLBOX#33\",\"source\":\"MYWALLBOX\",\"timestamp\":\"2024-02-06T20:08:28.757Z\"}}"
```

**Reply by kotope** [reply]
Thanks @jagheterfredrik for the information, the protocol does look quite simple.

However, this Redis framework is new for me, so need to look into it more deeply first to get more understanding. :-)
